### PR TITLE
[auth]: add authentication

### PR DESCRIPTION
**Pull request by legomushroom**
0. Adds the required Bearer auth strategy service API to the endpoints.
1. Adds the AAD JWT verification logic.
1. Adds the Cascade JWT verification logic.
2. Moves the extension to use the `@vs/vscode-account` module for authentication to share the access tokens with Live Share.

Still waits on:

0. Live Share change and release to reuse the auth strategies used by Live Share.

### CODE REVIEW

File: service/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
```
Comment (legomushroom) [inline service/.vscode/launch.json]: Remove.

File: service/.vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+    "version":"2.0.0",
```
Comment (legomushroom) [inline service/.vscode/tasks.json]: Remove.

File: service/mix.exs [hunk 3]
```diff
@@ -27,7 +27,9 @@ defmodule LiveShareCommunities.MixProject do
       {:plug_cowboy, "~> 2.0"},
       {:poison, "~> 4.0.1"},
       {:redix, ">= 0.0.0"},
-
+      {:joken, "~> 2.0-rc0"},
```
Comment (legomushroom) [inline service/mix.exs]: Remove `joken`.

File: src/api.ts [hunk 4]
```diff
@@ -1,80 +1,128 @@
 import axios from "axios";
+import { auth } from "./auth/auth";
 import { ICommunity } from "./store/model";
 
-const BASE_URL = "http://vslscommunitieswebapp.azurewebsites.net/v0";
+// const BASE_URL = "http://vslscommunitieswebapp.azurewebsites.net/v0";
+const BASE_URL = "http://localhost:4000/v0";
```
Comment (legomushroom) [inline src/api.ts]: Return the old one.

**Comment by arjun27** [comment]
looks awesome! would we want auth on the websocket?

also, couple of very minor things
- the dump.rdb file got committed (backed up from running Redis-server)
- can we remove package-lock.json (using yarn)?

File: .npmrc [hunk 5]
```diff
@@ -0,0 +1,2 @@
+@vs:registry=https://devdiv.pkgs.visualstudio.com/_packaging/VS/npm/registry/
```
Comment (arjun27) [inline .npmrc]: would this private feed need a PAT? how will this work with deploy.sh?

File: service/mix.exs (hunk 3 above)
Comment (arjun27) [reply service/mix.exs]: minor: removing dependency from mix.exs does not remove it from the lock file. that needs an additional `mix deps.unlock --unused`

File: .npmrc (hunk 5 above)
Comment (legomushroom) [reply .npmrc]: Yep, this means you will have to connect to the feed @VS feed at azure devops. If it deployed from the Azure devops it should be authenticated there. How do we deploy right now?

Comment (arjun27) [reply .npmrc]: at this stage I'm running `sh deploy.sh` locally. I'm also trying to get us access to the new GitHub Actions so we can set up that for CI/CD

Comment (legomushroom) [reply .npmrc]: If you run deploy locally and build the bits on your machine then it will work (if you will connect to the feed).

**Comment by legomushroom** [comment]
@arjun27 @lostintangent added the Cascade token validation, now the user will be able to use any (ms/gh) account thru live share.
Also addressed some feedback please take a look.


Re the WS auth, what are we using the WS for at the moment?

**Comment by arjun27** [comment]
> Re the WS auth, what are we using the WS for at the moment?

We're using the websocket transport that comes with [cowboy](https://github.com/ninenines/cowboy) (webserver for elixir) – cowboy is also listening for http. We could de-prioritize gating on websocket, at least till we implement private communities (#45)

Code for websocket, on the service, in case you want to review: https://github.com/vsls-contrib/communities/blob/master/service/lib/communities/websocket.ex

**Comment by legomushroom** [comment]
@arjun27 ya, but what scenarios we are using the WS for?

**Comment by arjun27** [comment]
> what scenarios we are using the WS for?

The websocket delivers real-time updates for: changes to the communities the user is a member of (member added/removed, founder changed), and new requests and messages in these communities

**Comment by legomushroom** [comment]
@arjun27 I see, thanks. Sounds like we do want to restrict those also. Maybe not in this PR though.

File: service/lib/communities/auth.ex [hunk 6]
```diff
@@ -0,0 +1,283 @@
+defmodule LiveShareCommunities.Authentication do
+  import Plug.Conn
+  use Memoize
+
+  def init(opts), do: opts
+
+  defmemo getKeys?(arg) do
```
Comment (lostintangent) [inline service/lib/communities/auth.ex]: Should we make all of these function names use snake case? (E.g. “get_keys?”). That seems to be the convention in Elixir.

File: src/config.ts [hunk 7]
```diff
@@ -18,6 +19,8 @@ export const config = {
         ConfigurationTarget.Global
       );
     }
+
+    await vscode.commands.executeCommand('liveshare.enable.vscode-account.auth');
```
Comment (lostintangent) [inline src/config.ts]: This file has been updated to import the `commands` API already, so you should pull the latest changes from `MASTER` and should be able to just call `commands.executeCommand('liveshare.enable.vscode-account.auth')`.

File: .vscode/launch.json [hunk 8]
```diff
@@ -3,34 +3,34 @@
...
-				"--extensionTestsPath=${workspaceFolder}/out/test/suite/index"
-			],
-			"outFiles": [
-				"${workspaceFolder}/out/test/**/*.js"
-			],
-			"preLaunchTask": "npm: watch"
-		}
-	]
-}
+    "version": "0.2.0",
```
Comment (lostintangent) [inline .vscode/launch.json]: If this is just formatting, can you revert your changes? We made some changes to this file in `MASTER`.

Comment (legomushroom) [reply .vscode/launch.json]: I took all the incoming, didnt make any changes to the file really.

File: src/config.ts (hunk 7 above)
Comment (legomushroom) [reply src/config.ts]: Done!

**Pull request merged**